zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationProvider.java

```diff
@@ -51,7 +57,18 @@
 public class X509AuthenticationProvider implements AuthenticationProvider {
 
     static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_SUPERUSER = "zookeeper.X509AuthenticationProvider.superUser";
-    private static final Logger LOG = LoggerFactory.getLogger(X509AuthenticationProvider.class);
+    /**
+     * The following System Property keys are used to extract clientId from the client cert.
+     * @see #getClientId(X509Certificate)
+     */
+    public static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_CLIENTCERTIDTYPE = "zookeeper.X509AuthenticationProvider.clientCertIdType";
+    public static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_CLIENTCERTIDSANMATCHTYPE = "zookeeper.X509AuthenticationProvider.clientCertIdSanMatchType";
+    // Match Regex is used to choose which entry to use
+    public static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_CLIENTCERTIDSANMATCHREGEX = "zookeeper.X509AuthenticationProvider.clientCertIdSanMatchRegex";
+    // Extract Regex is used to construct a client ID (acl entity) to return
+    public static final String ZOOKEEPER_X509AUTHENTICATIONPROVIDER_CLIENTCERTIDSANEXTRACTREGEX = "zookeeper.X509AuthenticationProvider.clientCertIdSanExtractRegex";
```

Comment:
Can these be set via zoo.cfg or are they set only via system properties?

Reply:
Both work. For zoo.cfg, please remove the `zookeeper.` prefix.